### PR TITLE
ln: Using unit test macros for ln and creating test cases for exception handling

### PR DESCRIPTION
**Pull request by ChrisDryden**
Going through the coverlay of ln to see where we are missing coverage and found a few edge cases that I was looking to cover with tests. It was interesting to see that one of the error messages was actually impossible to reach on unix, so I ended up adding a cfg to only run it for windows. 

I also updated the using of the macros in the tests since half of the tests were using them and half of them weren't

### CODE REVIEW

File: src/uu/ln/src/ln.rs [hunk 1]
```diff
@@ -132,7 +128,7 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
         verbose: matches.get_flag(options::VERBOSE),
     };
 
-    exec(&paths[..], &settings)
+    exec(&paths, &settings)
```
Comment (ChrisDryden) [inline src/uu/ln/src/ln.rs:131]: Was a bit confused when trying to understand what the [..] was doing when I was reading this, turns out its redundant

File: tests/by-util/test_ln.rs [hunk 2]
```diff
@@ -937,14 +883,37 @@ fn test_ln_non_utf8_paths() {
...
+    at.touch("c");
+    ucmd.args(&["-T", "a", "b", "c"])
+        .fails_with_code(1)
+        .stderr_contains("extra operand c")
+        .stderr_contains("--help");
+}
+
+#[cfg(target_os = "linux")]
+#[test]
+fn test_ln_cannot_stat_non_utf8() {
```
Comment (cakebaker) [inline tests/by-util/test_ln.rs]: I think this test is not correct. It matches the behavior of our implementation, however, if I do the steps manually with GNU `ln` (v9.9), I don't get an error:
```
$ mkdir target
$ touch $'file_\xFF'
ln -s $'file_\xFF' target
$ echo $?
0
$ ls -l target/
total 0
lrwxrwxrwx 1 dho dho 6 Dec 11 10:21 'file_'$'\377' -> 'file_'$'\377'
```
My suggestion is to add a todo, so we can fix it in a future PR.

Comment (ChrisDryden) [reply tests/by-util/test_ln.rs]: I think I was able to find the source of it. My understanding is that there was a to_str() that was was checked if None and that would only happen with non utf-8 filenames, by switching it to just calling filename directly, there's no optional. I made a commit that implements the fix for this too

**Comment by sylvestre** [comment]
fails with:

```
  Checking coreutils v0.5.0 (/home/runner/work/coreutils/coreutils)
  error: unused import: `uutests::util::TestScenario`
   --> tests/by-util/test_ln.rs:8:5
    |
  8 | use uutests::util::TestScenario;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D unused-imports` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(unused_imports)]`
  
  error: unused import: `util_name`
   --> tests/by-util/test_ln.rs:9:49
    |
  9 | use uutests::{at_and_ts, at_and_ucmd, new_ucmd, util_name};
    |                                                 ^^^^^^^^^
  
  error: could not compile `coreutils` (test "tests") due to 2 previous errors
Error: Final attempt failed. Child_process exited with error code 1

```
